chunkset_tpl.h

```diff
@@ -112,66 +67,83 @@ Z_INTERNAL uint8_t* CHUNKMEMSET(uint8_t *out, unsigned dist, unsigned len) {
 #ifdef HAVE_CHUNKMEMSET_2
     if (dist == 2) {
-        chunkmemset_2(from, &chunk);
+        chunkmemset_2(from, &chunk_load);
     } else
 #endif
 #ifdef HAVE_CHUNKMEMSET_4
     if (dist == 4) {
-        chunkmemset_4(from, &chunk);
+        chunkmemset_4(from, &chunk_load);
     } else
 #endif
 #ifdef HAVE_CHUNKMEMSET_8
     if (dist == 8) {
-        chunkmemset_8(from, &chunk);
+        chunkmemset_8(from, &chunk_load);
+    } else if (dist == sizeof(chunk_t)) {
+        loadchunk(from, &chunk_load);
     } else
 #endif
-    if (dist == sz) {
-        loadchunk(from, &chunk);
-    } else if (dist < sz) {
-        unsigned char *end = out + len - 1;
-        while (len > dist) {
-            out = CHUNKCOPY_SAFE(out, from, dist, end);
-            len -= dist;
+    {
+        /* This code takes string of length dist from "from" and repeats
+         * it for as many times as can fit in a chunk_t (vector register) */
+        int32_t cpy_dist;
+        int32_t bytes_remaining = sizeof(chunk_t);
+        uint8_t *cur_chunk = (uint8_t*)&chunk_load;
+        while (bytes_remaining) {
+            cpy_dist = MIN(dist, bytes_remaining);
+            memcpy(cur_chunk, from, cpy_dist);
+            bytes_remaining -= cpy_dist;
+            cur_chunk += cpy_dist;
         }
-        if (len > 0) {
-            out = CHUNKCOPY_SAFE(out, from, len, end);
+    }
+
+    /* If we're lucky enough and dist happens to be an even modulus of our vector length,
+     * we can do two stores per loop iteration, which for most ISAs, especially x86, is beneficial */
+    uint32_t chunk_mod = sizeof(chunk_t) % dist;
+    if (chunk_mod == 0) {
+        while (len >= (2 * sizeof(chunk_t))) {
+            storechunk(out, &chunk_load);
+            storechunk(out + sizeof(chunk_t), &chunk_load);
+            out += 2 * sizeof(chunk_t);
+            len -= 2 * sizeof(chunk_t);
+        }
+
+        while (len >= sizeof(chunk_t)) {
+            storechunk(out, &chunk_load);
+            out += sizeof(chunk_t);
+            len -= sizeof(chunk_t);
         }
-        return out;
     } else {
-        out = CHUNKUNROLL(out, &dist, &len);
-        return CHUNKCOPY(out, out - dist, len);
+        /* Since we don't have a "dist" length that divides evenly into a vector
+         * register, we can write the whole vector register but we need only
+         * advance by the amount of the whole string that fits in our chunk_t */
+        uint32_t adv_amount = sizeof(chunk_t) - chunk_mod;
+        while (len >= sizeof(chunk_t)) {
+            storechunk(out, &chunk_load);
+            len -= adv_amount;
+            out += adv_amount;
+        }
     }
 
-    unsigned rem = len % sz;
-    len -= rem;
     while (len) {
-        storechunk(out, &chunk);
-        out += sz;
-        len -= sz;
-    }
-
-    /* Last, deal with the case when LEN is not a multiple of SZ. */
-    if (rem) {
-        memcpy(out, from, rem);
-        out += rem;
+        int cpy_dist = MIN(dist, len);
```

Comment:
- Should we increase `dist += dist` at the end of the loop so the next time around it copies twice as much? 
- I assume at this point in the code that `len < sizeof(chunk_t)` correct? It can't be allowed to be greater because it would be greater than `sizeof(chunk_load)`.

Reply:
Actually this doesn't need to be a loop, I think we can just do a single iteration of size len (since len by definition is smaller than chunk_t).  The chunk should be loaded with pattern repeated as many times as will fit (and should have the head end of the one that doesn't).